Structured reasons for rejection

steventux opened the pull request:
## Context

We need to capture specific information about why a candidate's application is being rejected so that the feedback given is relevant and useful to the candidate.
<!-- Why are you making this change? What might surprise someone about it? -->

## Changes proposed in this pull request

Adds structured questions/reasons for rejecting an application in the provider interface and persists them in the same structure. 

### Initial questions with validation fail
![image](https://user-images.githubusercontent.com/93511/99249447-0db5dc80-2802-11eb-95d9-ae519adc6cfb.png)

### Other reasons for rejection

![image](https://user-images.githubusercontent.com/93511/99258544-a2bfd200-2810-11eb-8e73-35330eddc44b.png)


### Checking / Confirming answers
![image](https://user-images.githubusercontent.com/93511/99249246-b283ea00-2801-11eb-8698-c774bf539dad.png)

<!-- If there are UI changes, please include Before and After screenshots. -->

## Guidance to review

The bulk of the complexity is in the conditional validation logic, I've tried to keep this as simple as I can but there's a lot of it.

Does the feature flag strategy make sense in terms of releasing this feature and switching from the simpler single rejection reason?


<!-- How could someone else check this work? Which parts do you want more feedback on? -->

## Link to Trello card

https://trello.com/c/azXp4WN4/2238-%F0%9F%8F%88-structured-reasons-for-rejection

<!-- http://trello.com/123-example-card -->

## Things to check

- [x] This code does not rely on migrations in the same Pull Request
- [x] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [x] API release notes have been updated if necessary
- [x] New environment variables have been [added to the Azure config](https://github.com/DFE-Digital/apply-for-teacher-training#azure-hosting-devops-pipeline)

AgaDufrat commented on app/controllers/provider_interface/reasons_for_rejection_controller.rb:
```diff
@@ -0,0 +1,383 @@
+    end
+  end
+
+  class ReasonsForRejectionWizard
```
Would it be worth moving out of the controller to eg `app/forms/provider_interface`? 
Similarly the `ReasonsForRejection` class.

adamsilver commented on config/locales/reasons_for_rejection.yml:
```diff
@@ -0,0 +1,99 @@
+            candidate_behaviour_y_n:
+              blank: Select yes if it was due to candidate behaviour
+            candidate_behaviour_what_did_the_candidate_do:
+              blank: Select at least one reason related to the candidate's behaviour
```
```suggestion
              blank: Select at least one reason related to the candidate’s behaviour
```

adamsilver commented on config/locales/reasons_for_rejection.yml:
```diff
@@ -0,0 +1,99 @@
+            honesty_and_professionalism_y_n:
+              blank: Select yes if it was due to concerns about their honesty and professionalism
+            honesty_and_professionalism_concerns:
+              blank: Select at least one reason related to the candidate's honesty and professionalism
```
```suggestion
              blank: Select at least one reason related to the candidate’s honesty and professionalism
```

adamsilver commented on config/locales/reasons_for_rejection.yml:
```diff
@@ -0,0 +1,99 @@
+              blank: Enter details about safeguarding
+              too_long: The details you give about safeguarding must be 100 words or fewer
+            why_are_you_rejecting_this_application:
+              blank: Enter details about why you're rejecting their application
```
```suggestion
              blank: Enter details about why you’re rejecting their application
```

adamsilver commented on config/locales/reasons_for_rejection.yml:
```diff
@@ -0,0 +1,99 @@
+              blank: Enter any other advice or feedback
+              too_long: The advice or feedback you give must be 200 words or fewer
+            interested_in_future_applications_y_n:
+              blank: Select yes if you'd be interested in future applications from this candidate
```
```suggestion
              blank: Select yes if you’d be interested in future applications from this candidate
```

adamsilver commented:
@steventux the feedback is showing up buggy and the content isn’t quite right:

![image](https://user-images.githubusercontent.com/37163/99519350-ae82d400-2989-11eb-9f1d-972eb43327a4.png)

Should be like this:

![image](https://user-images.githubusercontent.com/37163/99519470-d7a36480-2989-11eb-9940-b53f8fbae6cc.png)

duncanjbrown commented:
@adamsilver most thorough design review in a PR ever? ⭐

mnacos commented on spec/system/provider_interface/reject_an_application_spec.rb:
```diff
@@ -0,0 +1,151 @@
+  end
+
+  def then_i_can_see_the_reasons_were_saved
+    expect(page).to have_content('Application rejected')
```
I imagine this step will include some more checks around the display of structured rejection reasons?

steventux replied:
@mnacos see https://github.com/DFE-Digital/apply-for-teacher-training/pull/3404/commits/4ca7143c010b08116efbc151bfd5e56083a35c45 for component to display reasons for rejection

steventux replied:
@adamsilver https://github.com/DFE-Digital/apply-for-teacher-training/commit/27396ca292b495ce36c024e37b4c28464d64a0a7 should fix the inset text problem when displaying reasons for rejection.

mnacos reviewed (commented):
Haven't fully reviewed this yet, but it seems to me we'll need a component for displaying structured reasons on application 'show'. Not sure yet if this should be re-used on the 'check step', because the 'show' component will also need to cope with unstructured reasons.

All this may be tackled in a different story, of course, as we have the feature flag.

adamsilver commented on config/locales/reasons_for_rejection.yml:
```diff
@@ -0,0 +1,99 @@
+        provider_interface/reasons_for_rejection_wizard:
+          attributes:
+            candidate_behaviour_y_n:
+              blank: Select yes if it was due to candidate behaviour
```
```suggestion
              blank: Select yes if it was related to candidate behaviour
```

adamsilver commented on config/locales/reasons_for_rejection.yml:
```diff
@@ -0,0 +1,99 @@
+              blank: Select at least one reason related to the candidate's behaviour
+            candidate_behaviour_other:
+              blank: Enter details about what the candidate did
+              too_long: The details you give about their behaviour must be 100 words or fewer
```
```suggestion
              too_long: Details about candidate behaviour must be 100 words or fewer
```

adamsilver commented on config/locales/reasons_for_rejection.yml:
```diff
@@ -0,0 +1,99 @@
+            candidate_behaviour_what_did_the_candidate_do:
+              blank: Select at least one reason related to the candidate's behaviour
+            candidate_behaviour_other:
+              blank: Enter details about what the candidate did
```
```suggestion
              blank: Enter details about candidate behaviour
```

adamsilver commented on config/locales/reasons_for_rejection.yml:
```diff
@@ -0,0 +1,99 @@
+              blank: Enter details about what the candidate did
+              too_long: The details you give about their behaviour must be 100 words or fewer
+            candidate_behaviour_what_to_improve:
+              blank: Enter details about how the candidate can improve their behaviour
```
```suggestion
              blank: Enter details about how candidate can improve their behaviour
```

adamsilver commented on config/locales/reasons_for_rejection.yml:
```diff
@@ -0,0 +1,99 @@
+              too_long: The details you give about their behaviour must be 100 words or fewer
+            candidate_behaviour_what_to_improve:
+              blank: Enter details about how the candidate can improve their behaviour
+              too_long: The details you give about how the candidate can improve their behaviour must be 100 words or fewer
```
```suggestion
              too_long: Details about how the candidate can improve their behaviour must be 100 words or fewer
```

adamsilver commented on config/locales/reasons_for_rejection.yml:
```diff
@@ -0,0 +1,99 @@
+              blank: Enter details about how the candidate can improve their behaviour
+              too_long: The details you give about how the candidate can improve their behaviour must be 100 words or fewer
+            quality_of_application_y_n:
+              blank: Select yes if it was due to the quality of their application
```
```suggestion
              blank: Select yes if it was related to the quality of their application
```

adamsilver commented on config/locales/reasons_for_rejection.yml:
```diff
@@ -0,0 +1,99 @@
+            quality_of_application_y_n:
+              blank: Select yes if it was due to the quality of their application
+            quality_of_application_which_parts_needed_improvement:
+              blank: Select at least one reason related to the quality of their application
```
```suggestion
              blank: Select at least one reason related to the quality of the candidate’s application
```

adamsilver commented on config/locales/reasons_for_rejection.yml:
```diff
@@ -0,0 +1,99 @@
+            quality_of_application_which_parts_needed_improvement:
+              blank: Select at least one reason related to the quality of their application
+            quality_of_application_personal_statement_what_to_improve:
+              blank: Enter details about how they can improve their personal statement
```
```suggestion
              blank: Enter details about how the candidate can improve their personal statement
```

adamsilver commented on config/locales/reasons_for_rejection.yml:
```diff
@@ -0,0 +1,99 @@
+              blank: Select at least one reason related to the quality of their application
+            quality_of_application_personal_statement_what_to_improve:
+              blank: Enter details about how they can improve their personal statement
+              too_long: The details you give about how they can improve their personal statement must be 100 words or fewer
```
```suggestion
              too_long: Details about how they can improve their personal statement must be 100 words or fewer
```

adamsilver commented on config/locales/reasons_for_rejection.yml:
```diff
@@ -0,0 +1,99 @@
+              too_long: The details you give about how they can improve their personal statement must be 100 words or fewer
+            quality_of_application_subject_knowledge_what_to_improve:
+              blank: Enter details about how they can improve their subject knowledge
+              too_long: The details you give about how the candidate can improve their subject knowledge must be 100 words or fewer
```
```suggestion
              too_long: Details about how the candidate can improve their subject knowledge must be 100 words or fewer
```

adamsilver commented on config/locales/reasons_for_rejection.yml:
```diff
@@ -0,0 +1,99 @@
+              blank: Enter details about how they can improve their subject knowledge
+              too_long: The details you give about how the candidate can improve their subject knowledge must be 100 words or fewer
+            quality_of_application_other_details:
+              blank: Enter details about the quality of the their application
```
```suggestion
              blank: Enter details about the quality of their application
```

adamsilver commented on config/locales/reasons_for_rejection.yml:
```diff
@@ -0,0 +1,99 @@
+              too_long: The details you give about how the candidate can improve their subject knowledge must be 100 words or fewer
+            quality_of_application_other_details:
+              blank: Enter details about the quality of the their application
+              too_long: The details you give about the quality of their application must be 100 words or fewer
```
```suggestion
              too_long: Details about the quality of their application must be 100 words or fewer
```

adamsilver commented on config/locales/reasons_for_rejection.yml:
```diff
@@ -0,0 +1,99 @@
+              too_long: The details you give about the quality of their application must be 100 words or fewer
+            quality_of_application_other_what_to_improve:
+              blank: Enter how they can improve their application
+              too_long: The details you give about how they can improve their application must be 100 words or fewer
```
```suggestion
              too_long: Details about how they can improve their application must be 100 words or fewer
```

adamsilver commented on config/locales/reasons_for_rejection.yml:
```diff
@@ -0,0 +1,99 @@
+              blank: Enter details about the quality of the their application
+              too_long: The details you give about the quality of their application must be 100 words or fewer
+            quality_of_application_other_what_to_improve:
+              blank: Enter how they can improve their application
```
```suggestion
              blank: Enter how they can improve the quality their application
```

adamsilver commented on config/locales/reasons_for_rejection.yml:
```diff
@@ -0,0 +1,99 @@
+              blank: Enter how they can improve their application
+              too_long: The details you give about how they can improve their application must be 100 words or fewer
+            qualifications_y_n:
+              blank: Select yes if it was due to their qualifications
```
```suggestion
              blank: Select yes if it was related to their qualifications
```

adamsilver commented on config/locales/reasons_for_rejection.yml:
```diff
@@ -0,0 +1,99 @@
+            qualifications_which_qualifications:
+              blank: Select at least one reason related to their qualifications
+            qualifications_other_details:
+              blank: Enter details about the issue with their qualifications
```
```suggestion
              blank: Enter details about the problem with their qualifications
```

adamsilver commented on config/locales/reasons_for_rejection.yml:
```diff
@@ -0,0 +1,99 @@
+              blank: Select at least one reason related to their qualifications
+            qualifications_other_details:
+              blank: Enter details about the issue with their qualifications
+              too_long: The details you give about their qualifications must be 100 words or fewer
```
```suggestion
              too_long: Details about their qualifications must be 100 words or fewer
```

adamsilver commented on config/locales/reasons_for_rejection.yml:
```diff
@@ -0,0 +1,99 @@
+              blank: Select yes if it was related to their performance at interview
+            performance_at_interview_what_to_improve:
+              blank: Enter details about how they can improve their performance at interview
+              too_long: The details you give about how they can improve their performance at interview must be 100 words or fewer
```
```suggestion
              too_long: Details about how they can improve their performance at interview must be 100 words or fewer
```

adamsilver commented on config/locales/reasons_for_rejection.yml:
```diff
@@ -0,0 +1,99 @@
+              blank: Select yes if it was because you offered them a place on another course
+            offered_on_another_course_details:
+              blank: Enter details about the place offered on another course
+              too_long: The details you give about an offer on another course must be 100 words or fewer
```
```suggestion
              too_long: Details about the place offered on another course must be 100 words or fewer
```

adamsilver commented on config/locales/reasons_for_rejection.yml:
```diff
@@ -0,0 +1,99 @@
+              blank: Enter details about the place offered on another course
+              too_long: The details you give about an offer on another course must be 100 words or fewer
+            honesty_and_professionalism_y_n:
+              blank: Select yes if it was due to concerns about their honesty and professionalism
```
```suggestion
              blank: Select yes if it was related to concerns about their honesty and professionalism
```

adamsilver commented on config/locales/reasons_for_rejection.yml:
```diff
@@ -0,0 +1,99 @@
+            honesty_and_professionalism_concerns:
+              blank: Select at least one reason related to the candidate's honesty and professionalism
+            honesty_and_professionalism_concerns_information_false_or_inaccurate_details:
+              blank: Enter details related to false or inaccurate information given
```
```suggestion
              blank: Enter details about the inaccurate or false information in the application
```

adamsilver commented on config/locales/reasons_for_rejection.yml:
```diff
@@ -0,0 +1,99 @@
+              blank: Select at least one reason related to the candidate's honesty and professionalism
+            honesty_and_professionalism_concerns_information_false_or_inaccurate_details:
+              blank: Enter details related to false or inaccurate information given
+              too_long: The details you give related to false or inaccurate information given must be 100 words or fewer
```
```suggestion
              too_long: Details about inaccurate or false information in the application must be 100 words or fewer
```

adamsilver commented on config/locales/reasons_for_rejection.yml:
```diff
@@ -0,0 +1,99 @@
+              blank: Enter details related to false or inaccurate information given
+              too_long: The details you give related to false or inaccurate information given must be 100 words or fewer
+            honesty_and_professionalism_concerns_plagiarism_details:
+              blank: Enter details related to evidence of plagiarism in personal statement or elsewhere
```
```suggestion
              blank: Enter details related to evidence of plagiarism in the application
```

adamsilver commented on config/locales/reasons_for_rejection.yml:
```diff
@@ -0,0 +1,99 @@
+              too_long: The details you give related to false or inaccurate information given must be 100 words or fewer
+            honesty_and_professionalism_concerns_plagiarism_details:
+              blank: Enter details related to evidence of plagiarism in personal statement or elsewhere
+              too_long: The details you give related to evidence of plagiarism in personal statement or elsewhere must be 100 words or fewer
```
```suggestion
              too_long: The details about evidence of plagiarism in the application must be 100 words or fewer
```

adamsilver commented on config/locales/reasons_for_rejection.yml:
```diff
@@ -0,0 +1,99 @@
+              blank: Enter details related to evidence of plagiarism in personal statement or elsewhere
+              too_long: The details you give related to evidence of plagiarism in personal statement or elsewhere must be 100 words or fewer
+            honesty_and_professionalism_concerns_references_details:
+              blank: Enter details related to references not supporting the application
```
```suggestion
              blank: Enter details about references not supporting the application
```

adamsilver commented on config/locales/reasons_for_rejection.yml:
```diff
@@ -0,0 +1,99 @@
+              too_long: The details you give related to evidence of plagiarism in personal statement or elsewhere must be 100 words or fewer
+            honesty_and_professionalism_concerns_references_details:
+              blank: Enter details related to references not supporting the application
+              too_long: The details you give related to references must be 100 words or fewer
```
```suggestion
              too_long: Details about references not supporting the application must be 100 words or fewer
```

adamsilver commented on config/locales/reasons_for_rejection.yml:
```diff
@@ -0,0 +1,99 @@
+              too_long: The details you give related to references must be 100 words or fewer
+            honesty_and_professionalism_concerns_other_details:
+              blank: Enter details about the candidate’s honesty and professionalism
+              too_long: The details you give related to their honesty and professionalism must be 100 words or fewer
```
```suggestion
              too_long: Details about the candidate’s honesty and professionalism must be 100 words or fewer
```

adamsilver commented on config/locales/reasons_for_rejection.yml:
```diff
@@ -0,0 +1,99 @@
+              blank: Select at least one reason related to safeguarding
+            safeguarding_concerns_candidate_disclosed_information_details:
+              blank: Enter details about the information they disclosed making them unsuitable to work with children
+              too_long: The details you give about the information they disclosed must be 100 words or fewer
```
```suggestion
              too_long: Details about the information the candidate disclosed must be 100 words or fewer
```

adamsilver commented on config/locales/reasons_for_rejection.yml:
```diff
@@ -0,0 +1,99 @@
+            safeguarding_concerns:
+              blank: Select at least one reason related to safeguarding
+            safeguarding_concerns_candidate_disclosed_information_details:
+              blank: Enter details about the information they disclosed making them unsuitable to work with children
```
```suggestion
              blank: Enter details about information the candidate disclosed which makes them unsuitable to work with children
```

adamsilver commented on config/locales/reasons_for_rejection.yml:
```diff
@@ -0,0 +1,99 @@
+              blank: Enter details about the information they disclosed making them unsuitable to work with children
+              too_long: The details you give about the information they disclosed must be 100 words or fewer
+            safeguarding_concerns_vetting_disclosed_information_details:
+              blank: Enter details about the information revealed by your vetting process
```
```suggestion
              blank: Enter details about the information found by your vetting process which makes the candidate unsuitable to work with children
```

adamsilver commented on config/locales/reasons_for_rejection.yml:
```diff
@@ -0,0 +1,99 @@
+              too_long: The details you give about the information they disclosed must be 100 words or fewer
+            safeguarding_concerns_vetting_disclosed_information_details:
+              blank: Enter details about the information revealed by your vetting process
+              too_long: The details you give about the information revealed by your vetting process must be 100 words or fewer
```
```suggestion
              too_long: Details about the information found by your vetting process which makes the candidate unsuitable to work with children must be 100 words or fewer
```

adamsilver commented on config/locales/reasons_for_rejection.yml:
```diff
@@ -0,0 +1,99 @@
+              too_long: The details you give about the information revealed by your vetting process must be 100 words or fewer
+            safeguarding_concerns_other_details:
+              blank: Enter details about safeguarding
+              too_long: The details you give about safeguarding must be 100 words or fewer
```
```suggestion
              too_long: Details about safeguarding must be 100 words or fewer
```

adamsilver commented on config/locales/reasons_for_rejection.yml:
```diff
@@ -0,0 +1,99 @@
+              too_long: The details you give about safeguarding must be 100 words or fewer
+            why_are_you_rejecting_this_application:
+              blank: Enter details about why you're rejecting their application
+              too_long: The details you give for rejecting this application must be 200 words or fewer
```
```suggestion
              too_long: Details about why you’re rejecting this application must be 200 words or fewer
```

adamsilver commented on config/locales/reasons_for_rejection.yml:
```diff
@@ -0,0 +1,99 @@
+            other_advice_or_feedback_y_n:
+              blank: Select yes if there is any other advice or feedback you‘d like to give
+            other_advice_or_feedback_details:
+              blank: Enter any other advice or feedback
```
```suggestion
              blank: Enter any other advice or feedback
```

adamsilver commented on config/locales/reasons_for_rejection.yml:
```diff
@@ -0,0 +1,99 @@
+              blank: Select yes if there is any other advice or feedback you‘d like to give
+            other_advice_or_feedback_details:
+              blank: Enter any other advice or feedback
+              too_long: The advice or feedback you give must be 200 words or fewer
```
```suggestion
              too_long: The other advice or feedback must be 200 words or fewer
```

adamsilver commented on app/views/provider_interface/reasons_for_rejection/edit_initial_questions.html.erb:
```diff
@@ -0,0 +1,185 @@
+        <%= f.govuk_radio_button :quality_of_application_y_n, 'No', label: { text: 'No' } %>
+      <% end %>
+
+      <%= f.govuk_radio_buttons_fieldset(:qualifications_y_n, legend: { size: 'm', text: 'Was it related to qualifications?' }) do %>
```
```suggestion
      <%= f.govuk_radio_buttons_fieldset(:qualifications_y_n, legend: { size: 'm', text: 'Was it related to their qualifications?' }) do %>
```

mnacos commented:
On the review app, I'm stuck on the commit page:

![image](https://user-images.githubusercontent.com/107591/100078797-d1a2fd00-2e3b-11eb-84b1-3110c0f9cc02.png)

mnacos commented:
Also, attempting to view a newly rejected application, (with safeguarding concenrs from the vetting process), I get a 500.

![image](https://user-images.githubusercontent.com/107591/100079526-a66cdd80-2e3c-11eb-9143-ede156db2a1e.png)

mnacos commented on app/components/provider_interface/reasons_for_rejection_component.html.erb:
```diff
@@ -0,0 +1,193 @@
+    <div class="app-feedback">
+      <%= content_tag(subheading_tag_name, 'Additional advice', class: 'govuk-heading-s') %>
+      <p class="govuk-body"><%= reasons_for_rejection.other_advice_or_feedback_details %></p>
+      <% if editable? %>
```
:+1:

mnacos commented on app/components/provider_interface/status_box_components/rejected_component.html.erb:
```diff
@@ -1,3 +1,13 @@
 <h2 class="govuk-heading-m govuk-!-font-size-27">Rejection details</h2>
 
-<%= render SummaryListComponent.new(rows: rejected_rows) %>
+<% if @application_choice.structured_rejection_reasons.present? %>
```
Great

mnacos reviewed (commented):
Thanks for including the presentation component within the application choice #show. I've flagged some weirdness on the review app (it may turn out to be the review app). Otherwise good.

duncanjbrown commented:
So nearly there! I checked "Would like them to apply again" but didn't see the "Future applications" section in the summary. And in the prototype the checkboxes are now big — is that correct @adamsilver ?

duncanjbrown commented:
Two more qs: 

- If I select "no" for either "safeguarding issues" or "honesty and professionalism" the go back and select "yes" for "safeguarding issues" should the summary still contain the "any other advice" field?
- If I click "cancel" on the last page should my answers so far be cleared?

steventux commented:
@duncanjbrown I'd assume the answer to your first question is no, if the step flow omits the 2nd step because of a change to the answers on the first then the summary shouldn't contain these. Another fun quirk of clearing out answers, @adamsilver can you confirm?
Second point: In the prototype cancel takes you back to the application but answers are retained, is this correct @adamsilver ?

steventux commented:
@duncanjbrown https://github.com/DFE-Digital/apply-for-teacher-training/pull/3404/commits/b67e0a275105af6e52d4052da3c9a5b7b1dbdee8 unsets other reasons if conditions change and these become irrelevant as per your question here https://github.com/DFE-Digital/apply-for-teacher-training/pull/3404#issuecomment-735712177 I'm assuming this is the right thing to do. I've addressed check box size and added the missing _Future applications_ section to the summary in the two commits before this.

duncanjbrown reviewed (approved):
LGTM!!! 🚀
